File: packages/toon-rpc-mcp/src/stdio.ts
/**
 * The MCP stdio transport: newline-delimited JSON-RPC messages on stdin and
 * stdout, one message per line with no embedded newlines. Anything else the
 * server prints must go to stderr.
 *
 * Requests are answered concurrently, as MCP allows, with at most
 * `maxPendingCalls` in flight; past that, and past a line longer than
 * `maxFrameBytes`, the stream is ended with a JSON-RPC error.
 */

import type { Readable, Writable } from 'node:stream';
import { DEFAULT_LIMITS } from '@reddb-io/toon-rpc';
import type { Limits } from '@reddb-io/toon-rpc';
import { McpServer } from './index.js';
import type { McpService } from './index.js';

export interface StdioOptions {
  input?: Readable;
  output?: Writable;
  limits?: Partial<Pick<Limits, 'maxFrameBytes' | 'maxPendingCalls'>>;
}

/** Serve one MCP session over stdio; resolves once input ends and every answer is written. */
export function serveStdio(service: McpService, options: StdioOptions = {}): Promise<void> {
  const input: Readable = options.input ?? process.stdin;
  const output: Writable = options.output ?? process.stdout;
  const maxLine = options.limits?.maxFrameBytes ?? DEFAULT_LIMITS.maxFrameBytes;
  const maxInFlight = options.limits?.maxPendingCalls ?? DEFAULT_LIMITS.maxPendingCalls;
  const server = new McpServer(service);
  const inFlight = new Set<Promise<void>>();
  let buffer = '';
  let stopped = false;

  const write = (line: string) => {
    if (!output.writableEnded) output.write(`${line}\n`);
  };
  const stop = (message: string) => {
    stopped = true;
    write(JSON.stringify({ jsonrpc: '2.0', id: null, error: { code: -32600, message } }));
    input.destroy();
  };
  const answer = (line: string) => {
    if (inFlight.size >= maxInFlight) return stop('Too many requests in flight');
    const pending: Promise<void> = server.handleLine(line).then((response) => {
      if (response !== undefined) write(response);
    });
    inFlight.add(pending);
    void pending.finally(() => inFlight.delete(pending));
  };

  return new Promise<void>((resolve) => {
    let finishing: Promise<void> | undefined;
    const finish = () =>
      (finishing ??= (async () => {
        while (inFlight.size > 0) await Promise.all(inFlight);
        output.end(resolve);
      })());
    input.setEncoding('utf8');
    input.on('data', (chunk: string) => {
      if (stopped) return;
      buffer += chunk;
      for (let end = buffer.indexOf('\n'); end !== -1; end = buffer.indexOf('\n')) {
        const line = buffer.slice(0, end).replace(/\r$/, '');
        buffer = buffer.slice(end + 1);
        if (line.trim() !== '') answer(line);
        if (stopped) return;
      }
      if (buffer.length > maxLine) stop('Message exceeds the size limit');
    });
    input.on('end', () => {
      if (!stopped && buffer.trim() !== '') answer(buffer);
      void finish();
    });
    input.on('close', () => void finish());
  });
}

File: crates/reddb-io-toon-rpc-mcp/src/stdio.rs
//! The MCP stdio transport: newline-delimited JSON-RPC messages, one per
//! line with no embedded newlines. Anything else the server prints must go to
//! stderr.

use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, BufReader};

use crate::{McpServer, McpService};

/// Serve one MCP session on this process's stdin and stdout until stdin ends.
pub async fn serve_stdio<S: McpService>(server: &McpServer<S>) -> std::io::Result<()> {
    serve_lines(
        server,
        tokio::io::stdin(),
        tokio::io::stdout(),
        16 * 1024 * 1024,
    )
    .await
}

/// Serve one MCP session over any line stream. A line longer than
/// `max_line_bytes` ends the session with a JSON-RPC error.
pub async fn serve_lines<S, R, W>(
    server: &McpServer<S>,
    input: R,
    mut output: W,
    max_line_bytes: usize,
) -> std::io::Result<()>
where
    S: McpService,
    R: AsyncRead + Unpin,
    W: AsyncWrite + Unpin,
{
    let mut input = BufReader::new(input);
    let mut line = Vec::new();
    loop {
        line.clear();
        let read = (&mut input)
            .take(max_line_bytes as u64 + 1)
            .read_until(b'\n', &mut line)
            .await?;
        if read == 0 {
            break;
        }
        if line.len() > max_line_bytes && line.last() != Some(&b'\n') {
            let refusal = r#"{"jsonrpc":"2.0","id":null,"error":{"code":-32600,"message":"Message exceeds the size limit"}}"#;
            output.write_all(refusal.as_bytes()).await?;
            output.write_all(b"\n").await?;
            break;
        }
        let text = String::from_utf8_lossy(&line);
        let text = text.trim_end_matches(['\n', '\r']);
        if text.trim().is_empty() {
            continue;
        }
        if let Some(answer) = server.handle_line(text) {
            output.write_all(answer.as_bytes()).await?;
            output.write_all(b"\n").await?;
            output.flush().await?;
        }
    }
    output.shutdown().await
}
